xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,724 @@
+    private final AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub;
+
+    private StreamObserver<DiscoveryRequest> requestWriter;
+    private boolean responseReceived;
+    private boolean closed;
+
+    // Last successfully applied version_info for each resource type. Starts with empty string.
+    // A version_info is used to update management server with client's most recent knowledge of
+    // resources.
+    private String ldsVersion = "";
+    private String rdsVersion = "";
+
+    // Response nonce for the most recently received discovery responses of each resource type.
+    // Client initiated requests start response nonce with empty string.
+    // A nonce is used to indicate the specific DiscoveryResponse each DiscoveryRequest
+    // corresponds to.
+    // A nonce becomes stale following a newer nonce being presented to the client in a
+    // DiscoveryResponse.
+    private String ldsRespNonce = "";
+    private String rdsRespNonce = "";
+
+    // Most recently requested resource name for each resource type. Note the resource_name in
+    // LDS requests will always be "xds:" URI (including port suffix if present).
+    private String rdsResourceName = "";
+
+    // A prepared ACK request. Every DiscoveryResponse is replied with a DiscoveryRequest as ACK
+    // (with version_info set to response's version_info) or NACK (with version_info set to last
+    // ACKed response's version_info).
+    @Nullable
+    private DiscoveryRequest pendingAckRequest;
+
+    private AdsStream(AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub) {
+      this.stub = checkNotNull(stub, "stub");
+    }
+
+    private void start() {
+      requestWriter = stub.withWaitForReady().streamAggregatedResources(this);
+    }
+
+    @Override
+    public void onNext(final DiscoveryResponse response) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          responseReceived = true;
+          String typeUrl = response.getTypeUrl();
+          if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+            handleLdsResponse(response);
+          } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+            handleRdsResponse(response);
+          }
+          // TODO(zdapeng): add CDS/EDS response handles.
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
```

Comment:
Shouldn't this be really a method of `XdsClientImpl` instead of `AdsStream`? Or you can split this into stuff to do inside `AdsStream` and most of the rest in `XdsClientImpl` but that also doesn't seem necessary. The current code creates a new `adsStream` instance but still runs on the old instance without really using any instance values from the old instance so usage like `adsStream.sendLdsRequest(ldsResourceName);` is quite confusing.

Reply:
The relation between `XdsClientImpl` and `AdsStream` is __hasA__. Idea borrowed from [Grpclb](https://github.com/grpc/grpc-java/blob/6d3fb531285628f4ca7c8b281dabcddc0db2efc3/grpclb/src/main/java/io/grpc/grpclb/GrpclbState.java#L634).